File: backend/app/api/v1/endpoints/schedules.py
from typing import List
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from app.api.deps import get_db, get_current_user, get_current_director
from app.models.user import User, UserRole
from app.models.schedule import Schedule, DayOfWeek
from app.models.student import StudentProfile
from app.models.teacher import TeacherProfile
from app.schemas.schedule import ScheduleCreate, ScheduleRead, ScheduleUpdate

router = APIRouter()


@router.get("", response_model=List[ScheduleRead])
async def get_schedules(
    class_id: int = None,
    teacher_id: int = None,
    day_of_week: DayOfWeek = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get all schedules with optional filters."""
    query = db.query(Schedule)

    if class_id:
        query = query.filter(Schedule.class_id == class_id)
    if teacher_id:
        query = query.filter(Schedule.teacher_id == teacher_id)
    if day_of_week:
        query = query.filter(Schedule.day_of_week == day_of_week)

    # Filter by active schedules
    today = date.today()
    query = query.filter(Schedule.effective_from <= today)
    query = query.filter(
        (Schedule.effective_to.is_(None)) | (Schedule.effective_to >= today)
    )

    schedules = query.order_by(Schedule.day_of_week, Schedule.start_time).all()
    return schedules


@router.get("/my", response_model=List[ScheduleRead])
async def get_my_schedule(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get current user's schedule (student's class or teacher's classes)."""
    today = date.today()
    query = db.query(Schedule)

    if current_user.role == UserRole.STUDENT:
        # Get student's class schedule
        student_profile = (
            db.query(StudentProfile)
            .filter(StudentProfile.user_id == current_user.id)
            .first()
        )
        if not student_profile or not student_profile.class_id:
            return []
        query = query.filter(Schedule.class_id == student_profile.class_id)

    elif current_user.role == UserRole.TEACHER:
        # Get teacher's schedule
        teacher_profile = (
            db.query(TeacherProfile)
            .filter(TeacherProfile.user_id == current_user.id)
            .first()
        )
        if not teacher_profile:
            return []
        query = query.filter(Schedule.teacher_id == teacher_profile.id)

    else:
        # Directors/admins see all
        pass

    # Filter by active schedules
    query = query.filter(Schedule.effective_from <= today)
    query = query.filter(
        (Schedule.effective_to.is_(None)) | (Schedule.effective_to >= today)
    )

    schedules = query.order_by(Schedule.day_of_week, Schedule.start_time).all()
    return schedules


@router.post("", response_model=ScheduleRead, status_code=status.HTTP_201_CREATED)
async def create_schedule(
    schedule_in: ScheduleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_director),
):
    """Create a new schedule entry (Director only)."""
    # Check for conflicts
    existing = (
        db.query(Schedule)
        .filter(
            Schedule.class_id == schedule_in.class_id,
            Schedule.day_of_week == schedule_in.day_of_week,
            Schedule.start_time == schedule_in.start_time,
        )
        .first()
    )
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Schedule conflict: class already has a lesson at this time",
        )

    schedule = Schedule(**schedule_in.model_dump())
    db.add(schedule)
    db.commit()
    db.refresh(schedule)
    return schedule


@router.put("/{schedule_id}", response_model=ScheduleRead)
async def update_schedule(
    schedule_id: int,
    schedule_in: ScheduleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_director),
):
    """Update a schedule entry (Director only)."""
    schedule = db.query(Schedule).filter(Schedule.id == schedule_id).first()
    if not schedule:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Schedule not found",
        )

    update_data = schedule_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(schedule, field, value)

    db.commit()
    db.refresh(schedule)
    return schedule


@router.delete("/{schedule_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_schedule(
    schedule_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_director),
):
    """Delete a schedule entry (Director only)."""
    schedule = db.query(Schedule).filter(Schedule.id == schedule_id).first()
    if not schedule:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Schedule not found",
        )

    db.delete(schedule)
    db.commit()
    return None

File: backend/src/routes/v1/schedules.ts
import type { DayOfWeek } from '@prisma/client';
import type { FastifyPluginAsync } from 'fastify';

import { DIRECTOR_PLUS_ROLES } from '../../lib/auth.js';
import { conflict, notFound } from '../../lib/errors.js';
import { prisma } from '../../lib/prisma.js';
import { serializeSchedule } from '../../lib/serializers.js';
import { parseDateOnly, parseTimeOnly } from '../../lib/time.js';

const schedulesRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get('', { preHandler: [fastify.authenticate] }, async (request) => {
    const query = request.query as {
      class_id?: string | number;
      teacher_id?: string | number;
      day_of_week?: DayOfWeek;
    };

    const today = new Date();
    const startOfToday = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), today.getUTCDate()));

    const schedules = await prisma.schedule.findMany({
      where: {
        ...(query.class_id ? { class_id: Number(query.class_id) } : {}),
        ...(query.teacher_id ? { teacher_id: Number(query.teacher_id) } : {}),
        ...(query.day_of_week ? { day_of_week: query.day_of_week } : {}),
        effective_from: { lte: startOfToday },
        OR: [{ effective_to: null }, { effective_to: { gte: startOfToday } }],
      },
      orderBy: [{ day_of_week: 'asc' }, { start_time: 'asc' }],
    });

    return schedules.map(serializeSchedule);
  });

  fastify.get('/my', { preHandler: [fastify.authenticate] }, async (request) => {
    const currentUser = request.currentUser!;

    const today = new Date();
    const startOfToday = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), today.getUTCDate()));

    let where: Record<string, unknown> = {
      effective_from: { lte: startOfToday },
      OR: [{ effective_to: null }, { effective_to: { gte: startOfToday } }],
    };

    if (currentUser.role === 'student') {
      const student = await prisma.studentProfile.findUnique({ where: { user_id: currentUser.id } });
      if (!student || !student.class_id) {
        return [];
      }
      where = { ...where, class_id: student.class_id };
    } else if (currentUser.role === 'teacher') {
      const teacher = await prisma.teacherProfile.findUnique({ where: { user_id: currentUser.id } });
      if (!teacher) {
        return [];
      }
      where = { ...where, teacher_id: teacher.id };
    }

    const schedules = await prisma.schedule.findMany({
      where,
      orderBy: [{ day_of_week: 'asc' }, { start_time: 'asc' }],
    });

    return schedules.map(serializeSchedule);
  });

  fastify.post('', { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] }, async (request, reply) => {
    const body = request.body as {
      class_id: number;
      subject_id: number;
      teacher_id: number;
      day_of_week: DayOfWeek;
      start_time: string;
      end_time: string;
      room?: string | null;
      effective_from: string;
      effective_to?: string | null;
    };

    const existing = await prisma.schedule.findFirst({
      where: {
        class_id: body.class_id,
        day_of_week: body.day_of_week,
        start_time: parseTimeOnly(body.start_time),
      },
    });

    if (existing) {
      conflict('Schedule conflict: class already has a lesson at this time');
    }

    const created = await prisma.schedule.create({
      data: {
        class_id: body.class_id,
        subject_id: body.subject_id,
        teacher_id: body.teacher_id,
        day_of_week: body.day_of_week,
        start_time: parseTimeOnly(body.start_time),
        end_time: parseTimeOnly(body.end_time),
        room: body.room ?? null,
        effective_from: parseDateOnly(body.effective_from),
        effective_to: body.effective_to ? parseDateOnly(body.effective_to) : null,
      },
    });

    return reply.status(201).send(serializeSchedule(created));
  });

  fastify.put(
    '/:schedule_id',
    { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] },
    async (request) => {
      const params = request.params as { schedule_id: string };
      const body = request.body as {
        class_id?: number;
        subject_id?: number;
        teacher_id?: number;
        day_of_week?: DayOfWeek;
        start_time?: string;
        end_time?: string;
        room?: string | null;
        effective_from?: string;
        effective_to?: string | null;
      };

      const scheduleId = Number(params.schedule_id);
      const existing = await prisma.schedule.findUnique({ where: { id: scheduleId } });
      if (!existing) {
        notFound('Schedule not found');
      }

      const updated = await prisma.schedule.update({
        where: { id: scheduleId },
        data: {
          ...(body.class_id !== undefined ? { class_id: body.class_id } : {}),
          ...(body.subject_id !== undefined ? { subject_id: body.subject_id } : {}),
          ...(body.teacher_id !== undefined ? { teacher_id: body.teacher_id } : {}),
          ...(body.day_of_week !== undefined ? { day_of_week: body.day_of_week } : {}),
          ...(body.start_time !== undefined ? { start_time: parseTimeOnly(body.start_time) } : {}),
          ...(body.end_time !== undefined ? { end_time: parseTimeOnly(body.end_time) } : {}),
          ...(body.room !== undefined ? { room: body.room } : {}),
          ...(body.effective_from !== undefined ? { effective_from: parseDateOnly(body.effective_from) } : {}),
          ...(body.effective_to !== undefined
            ? { effective_to: body.effective_to ? parseDateOnly(body.effective_to) : null }
            : {}),
        },
      });

      return serializeSchedule(updated);
    }
  );

  fastify.delete(
    '/:schedule_id',
    { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] },
    async (request, reply) => {
      const params = request.params as { schedule_id: string };
      const scheduleId = Number(params.schedule_id);

      const existing = await prisma.schedule.findUnique({ where: { id: scheduleId } });
      if (!existing) {
        notFound('Schedule not found');
      }

      await prisma.schedule.delete({ where: { id: scheduleId } });
      return reply.status(204).send();
    }
  );
};

export default schedulesRoutes;
